### PR TITLE
Handle generic maps and channels

### PR DESCRIPTION
**Pull request by ndtretyak**
## What was changed
`determinism` checker now unwraps generic types, so that functions iterating over a generic `map` are marked non-deterministic

### CODE REVIEW

File: contrib/tools/workflowcheck/determinism/checker.go [hunk 1]
```diff
@@ -381,6 +366,41 @@ func (c *collector) collectFuncInfo(fn *types.Func, decl *ast.FuncDecl) {
 	})
 }
 
+func (c *collector) checkRangeType(rangeType types.Type, n ast.Node, fn *types.Func) Reason {
+L: // Unwrap named or generic type
```
Comment (Quinn-With-Two-Ns) [inline contrib/tools/workflowcheck/determinism/checker.go:370]: `L:`?

Comment (ndtretyak) [reply contrib/tools/workflowcheck/determinism/checker.go:370]: Should I change it to `Loop:`?

Comment (Quinn-With-Two-Ns) [reply contrib/tools/workflowcheck/determinism/checker.go:370]: Originally I was confused why you added a  label at all. I see  you changed the if-else to a select so it is needed now.

Bit of a nitpick, but I'd prefer the label be `Loop`

Comment (ndtretyak) [reply contrib/tools/workflowcheck/determinism/checker.go]: fixed

File: contrib/tools/workflowcheck/determinism/checker.go [hunk 2]
```diff
@@ -381,6 +366,42 @@ func (c *collector) collectFuncInfo(fn *types.Func, decl *ast.FuncDecl) {
...
+	// Unwrap named or generic type
+	for {
+		switch rangeType.(type) {
+		case *types.Named, *types.TypeParam:
+			rangeType = rangeType.Underlying()
+		default:
+			break Loop
+		}
+	}
+	switch t := rangeType.(type) {
```
Comment (cretz) [inline contrib/tools/workflowcheck/determinism/checker.go:380]:
```suggestion
	switch t := rangeType.(type) {
	case *types.Named:
		return c.checkRangeType(rangeType.Underlying(), n, fn)
	case *types.TypeParam:
		return c.checkRangeType(rangeType.Underlying(), n, fn)
```

Recursion reads a bit better to me, but not a big deal

**Review by cretz — APPROVED** [review]
Looks great, thanks for the contribution!

**Pull request merged**